### ISSUE TITLE
Allow to use refresh=false when using the CLI driven workflow

### ISSUE THREAD

**Issue by igorbrites** [AzBuilder/terrakube#647]
### Bug description 🐞

I'm moving all my workspaces from Terraform Cloud, and created all the global variables from there, including AWS credentials (access and secret keys). Now I need to run a plan for my EKS cluster, but it seems like Terrakube is not using the credentials I'm passing to it, showing the error below when it try to touch secrets and configmaps inside the cluster (btw it's not the same cluster where terrakube is running on):

```
Error: secrets "dev-certificate" is forbidden: User "system:serviceaccount:terrakube:default" cannot get resource "secrets" in API group "" in the namespace "dev"

Error: configmaps "aws-auth" is forbidden: User "system:serviceaccount:terrakube:default" cannot get resource "configmaps" in API group "" in the namespace "kube-system"
```

### Steps to reproduce

AWS Credentials as global variables:
![image](https://github.com/AzBuilder/terrakube/assets/1479917/c95af349-6231-41fa-aabd-04890316177a)

The following Terraform code:

```hcl
provider "aws" {
  region = "us-east-1"
}

data "aws_eks_cluster" "cluster" {
  name = "my-cluster"
}

data "aws_eks_cluster_auth" "cluster" {
  name = "my-cluster"
}

provider "kubernetes" {
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority[0].data)
  host                   = data.aws_eks_cluster.cluster.endpoint
  token                  = data.aws_eks_cluster_auth.cluster.token
}

resource "kubernetes_secret_v1" "dev_certificate" {
  metadata {
    name      = "dev-certificate"
    namespace = "dev"
  }

  data = {
    "secret"         = "value"
    "another-secret" = "another-value"
  }
}
```

### Expected behavior

Terrakube should use the given credentials and don't use its own ServiceAccount.

### Anything else?

Using this space for something else: how should I parallelize the runs? Just raise more executor pods?

**Comment by igorbrites** [comment]
Just as FYI, this is happening with every workspace that has Kubernetes resources.

**Comment by igorbrites** [comment]
Another this I'm confused with is that the error messages are getting mashed up with the plan:

```
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place

���
��� Error: reading RDS Cluster (my-other-cluster): empty result
��� 
���   with data.aws_rds_cluster.my_other_cluster,
���   on main.tf line 5, in data "aws_rds_cluster" "my_other_cluster":
���    5: data "aws_rds_cluster" "my_other_cluster" {
Terraform planned the following actions, but then encountered a problem:
��� 

  # module.db_starter.aws_iam_role.this[0] will be updated in-place
���
  ~ resource "aws_iam_role" "this" {
���
��� Error: reading RDS Cluster (my-cluster): empty result
      ~ assume_role_policy    = jsonencode(
��� 
���   with data.aws_rds_cluster.my_cluster,
���   on main.tf line 9, in data "aws_rds_cluster" "my_cluster":
          ~ {
              ~ Statement = [
���    9: data "aws_rds_cluster" "my_cluster" {
                  ~ {
��� 
                      ~ Condition = {
���
                          ~ StringEquals = {
```

**Comment by alfespa17** [comment]
Hello @igorbrites 

Firts to answer this: _"Using this space for something else: how should I parallelize the runs? Just raise more executor pods?"_, one executor pod can handle several terraform cli process in parallel, if you want you can add executor pods you can modify the replica value when you deploy the helm chart.

**Comment by alfespa17** [comment]
For the other issue I am not really sure it looks like it is trying to use a service account to connect.... 

when you deploy the helm chart did you include a "service account" for the executor pod?

**Comment by igorbrites** [comment]
@alfespa17 no I don't, it's the default ServiceAccount for the namespace, it gets created automatically. What I'm not sure is, if I'm passing the credentials, why it's trying to do it without them?

**Comment by igorbrites** [comment]
> For the other issue I am not really sure it looks like it is trying to use a service account to connect....
> 
> when you deploy the helm chart did you include a "service account" for the executor pod?

Another thing is, if I run locally, passing the same vars, it works.

**Comment by alfespa17** [comment]
You are passing the aws credentials correctly, basically the global variables AWS_ACCESS_KEY_ID and  AWS_SECRET_ACCESS_KEY are passed to the executor, you could check in the API pods to see all the values that are sent to the executor.

Internally the executor component is just creating a terraform CLI process and is passing the environment variables to that particular process...... so I am not really sure....

But if you run that locally I guess you have some kubeconfig file in your computer... maybe it is working because it is using your current kubeconfig confirguration..... could you remove your kubeconfig configuration temporarly and check if it continue working locally just using the environment variables AWS_ACCESS_KEY_ID and  AWS_SECRET_ACCESS_KEY?

**Comment by igorbrites** [comment]
> You are passing the aws credentials correctly, basically the global variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are passed to the executor, you could check in the API pods to see all the values that are sent to the executor.
> 
> Internally the executor component is just creating a terraform CLI process and is passing the environment variables to that particular process...... so I am not really sure....
> 
> But if you run that locally I guess you have some kubeconfig file in your computer... maybe it is working because it is using your current kubeconfig confirguration..... could you remove your kubeconfig configuration temporarly and check if it continue working locally just using the environment variables AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY?

Let me try, I'll let you know. Thanks!

**Comment by igorbrites** [comment]
While checking the errors, I've found some messages that may help us:

```
Error: reading Route 53 Record (<<REDACTED>>_*.<<REDACTED>>_A): AccessDenied: User: arn:aws:iam::<<REDACTED>>:user/terraform is not authorized to access this resource
	status code: 403, request id: <<REDACTED>>

Error: failed to create kubernetes rest client for read of resource: Get "http://localhost/api?timeout=32s": dial tcp [::1]:80: connect: connection refused
```

That Route53 is in one account, though the terraform is on another. To make this work, I pass the assumable role directly to the AWS provider, so the user's access key can assume that role, and terraform can apply things on the other account.

```hcl
provider "aws" {
  region = "us-east-1"
  assume_role {
    role_arn = var.aws_role_arn
  }
}
```

So this makes me think that the provider is not assuming the one in this case.

Now the other error message tells me that the aws_eks_cluster data source didn't get the right credentials (for the same issue above), so terraform is trying to consume the local Kubernetes API, so that's the reason we are seeing the terrakube ServiceAccount, it's trying to find stuff on the same cluster where it's installed.

I'll try to simulate the assume role and let you know.

**Comment by igorbrites** [comment]
based on these messages, I see you set up the S3 backend in the background:

```
terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Initializing the backend...
terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - 
terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Successfully configured the backend "s3"! Terraform will automatically
terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - use this backend unless the backend configuration changes.
```

Maybe this config is overwriting the credentials for the provider?

**Comment by alfespa17** [comment]
> based on these messages, I see you set up the S3 backend in the background:
> 
> ```
> terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Initializing the backend...
> terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - 
> terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Successfully configured the backend "s3"! Terraform will automatically
> terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - use this backend unless the backend configuration changes.
> ```
> 
> Maybe this config is overwriting the credentials for the provider?

That one is just to override the backend state configuration, so terrakube can use the s3, azure or gcp backen storage to save the terraform state.

Example reference:
https://github.com/AzBuilder/terrakube/blob/2bb629eb16235b8fd9dc47e93e41a490fb0a1ff2/executor/src/main/java/org/terrakube/executor/plugin/tfstate/aws/AwsTerraformStateImpl.java#L70

**Comment by igorbrites** [comment]
> > based on these messages, I see you set up the S3 backend in the background:
> > ```
> > terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Initializing the backend...
> > terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - 
> > terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - Successfully configured the backend "s3"! Terraform will automatically
> > terrakube-executor-85b478cd5b-slxpg terrakube-executor [ForkJoinPool-1-worker-5] INFO org.terrakube.executor.service.terraform.TerraformExecutorServiceImpl - use this backend unless the backend configuration changes.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe this config is overwriting the credentials for the provider?
> 
> That one is just to override the backend state configuration, so terrakube can use the s3, azure or gcp backen storage to save the terraform state.
> 
> Example reference:
> 
> https://github.com/AzBuilder/terrakube/blob/2bb629eb16235b8fd9dc47e93e41a490fb0a1ff2/executor/src/main/java/org/terrakube/executor/plugin/tfstate/aws/AwsTerraformStateImpl.java#L70

Hmmm ok, got it, so it shouldn't impact anything.

**Comment by igorbrites** [comment]
Some updates:

- I've created all the env vars on an isolated shell, without aws profiles nor kubeconfigs, and ran `aws sts get-caller-identity` just to be sure it was using the right user, then I changed the workspace to run locally. If I run only `terraform plan`, I see a similar error:

```
╷
│ Error: Get "http://localhost/api/v1/namespaces/dev/secrets/dev-certificate": dial tcp [::1]:80: connect: connection refused
│ 
│   with kubernetes_secret_v1.dev_certificate,
│   on cluster.tf line 81, in resource "kubernetes_secret_v1" "dev_certificate":
│   81: resource "kubernetes_secret_v1" "dev_certificate" {
│ 
╵
╷
│ Error: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp [::1]:80: connect: connection refused
│ 
│   with module.eks_cluster.module.eks.kubernetes_config_map_v1_data.aws_auth[0],
│   on .terraform/modules/eks_cluster.eks/main.tf line 562, in resource "kubernetes_config_map_v1_data" "aws_auth":
│  562: resource "kubernetes_config_map_v1_data" "aws_auth" {
│ 
╵
```

- Though if I pass `-refresh=false` to the plan, it runs successfully.
- Then I changed the workspace to run remotely again, and tested the `-refresh=false`, but I still see the same error:

```
Error: secrets "dev-otc-certificate" is forbidden: User "system:serviceaccount:terrakube:default" cannot get resource "secrets" in API group "" in the namespace "dev"
```

So it's definitely an issue with the Kubernetes provider that I remember facing before, that's how I came with the `-refresh=false` idea. There's an [open issue there since 2020 about this problem](https://github.com/hashicorp/terraform-provider-kubernetes/issues/1028).

But this shows me that the `-refresh=false` is not being sent to the remote terraform CLI. How do I make this working? I saw there's a flow feature, but not sure if it applies for this issue.

**Comment by alfespa17** [comment]
Hello @igorbrites right now there is no way to add the refresh flag when running the terraform plan.

We maintain this library, we will have to add that feature in the terraform library before using it in Terrakube.

https://github.com/AzBuilder/terraform-spring-boot

**Comment by igorbrites** [comment]
Hmmm ok, got it. For now, I'll use local runs for the Kubernetes workspaces. I'll keep this issue open until we get it implemented (I could open a PR there, but I know nothing about Java 😅).

Regardless, thanks for all the support today @alfespa17! Much appreciated!

**Comment by igorbrites** [comment]
@alfespa17 I see you already implemented the refresh flag, thanks! Waiting for the 2.19.0 😄.

But I'm seeing this error on some workspaces now:

```
ERROR 1 --- [nio-8080-exec-1] o.t.a.p.s.aws.AwsStorageTypeServiceImpl  : S3 Not found: The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID:<<REDACTED>>; S3 Extended Request ID: 3FBpS5M<<REDACTED>>ED11ZgBMI=; Proxy: null)
```

I only see this during applies, and when this happens, the apply tries to delete all the resources. Plans look right.

At least Terraform understands that something wrong is going on, and prevents this from being applied:

```
***************************************
Running Terraform APPLY
***************************************
���
��� Error: Inconsistent dependency lock file
��� 
��� The given plan file was created with a different set of external dependency
��� selections than the current configuration. A saved plan can be applied only
��� to the same configuration it was created from.
��� 
��� Create a new plan from the updated configuration.
���
```

Have you ever seen this kind of error? How could I fix this?

**Comment by igorbrites** [comment]
Actually it's the same version, I'm applying it using github actions, and I fixed that version because is the last one before their license changing:

```yaml
- name: Setup Terraform
  uses: hashicorp/setup-terraform@v2
  with:
    terraform_version: '1.5.7'
```

And this is the workspace:
![image](https://github.com/AzBuilder/terrakube/assets/1479917/16482221-0344-401e-bc78-943bf316a6dd)

**Comment by igorbrites** [comment]
Idk if it helps, but here's the whole context for that error:

```log
INFO 1  --- [nio-8080-exec-9] o.t.a.p.s.c.TerraformOutputController    : Reading output from storage
INFO 1  --- [nio-8080-exec-9] o.t.a.p.s.aws.AwsStorageTypeServiceImpl  : Searching: tfoutput/6ccdd710-<<REDACTED>>d18507/541/4c95378d-<<REDACTED>>-b2794ee6c434.tfoutput
INFO 1  --- [nio-8080-exec-1] o.t.a.p.s.aws.AwsStorageTypeServiceImpl  : Searching: tfstate/6ccdd710-<<REDACTED>>d18507/3125d278-<<REDACTED>>-04d3bbbfaab9/state/97c9fc18-<<REDACTED>>-26eca05ef469.json
INFO 1  --- [nio-8080-exec-7] o.t.a.p.s.aws.AwsStorageTypeServiceImpl  : Searching: /tfoutput/context/541/context.json
ERROR 1 --- [nio-8080-exec-7] o.t.a.p.s.aws.AwsStorageTypeServiceImpl  : S3 Not found: The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: <<REDACTED>>; S3 Extended Request ID: 3k5RB8RVisZv7Skloa3AX<<REDACTED>>A+iOGL4M=; Proxy: null)
```

**Comment by igorbrites** [comment]
I see the method that tries to get that context:

https://github.com/AzBuilder/terrakube/blob/2bb629eb16235b8fd9dc47e93e41a490fb0a1ff2/api/src/main/java/org/terrakube/api/plugin/storage/azure/AzureStorageTypeServiceImpl.java#L130-L138

So the bucket should have a context folder. What do you think that could prevent the folder creation?

**Comment by alfespa17** [comment]
There is a feature in Terrakube that you can use to save a "[context](https://docs.terrakube.io/user-guide/organizations/templates/persistent-context)" for a job execution for example save the the cost when integrating with infracost and print the values in the UI, you don't need to worry about that because you are not using that particular feature for now

**Comment by alfespa17** [comment]
You main issue is this one ***"Inconsistent dependency lock file"***, maybe you can try doing a terraform pull to download the state locally, after that you can run the "terraform init  -upgrade" and do a terraform push to update the state in Terrakube

**Comment by igorbrites** [comment]
> You main issue is this one _**"Inconsistent dependency lock file"**_, maybe you can try doing a terraform pull to download the state locally, after that you can run the "terraform init -upgrade" and do a terraform push to update the state in Terrakube

But this error happened after it tried to destroy everything, my real concern is that the apply tried to remove all the resources.

And I see that some runs got stuck on the approval phase, but if I click to Discard them, I see a 403 on the console with the message `{"errors":[{"detail":"UpdatePermission Denied"}]}`, and the S3 error on the API logs. Though I've created a team with all access.

**Comment by alfespa17** [comment]
What is happening is that there is some special "[template](https://docs.terrakube.io/user-guide/organizations/templates/default-templates#terraform-plan-apply-cli-template)" for the CLI-driven workflow, it has some "special" team that is used internally when you answer "yes" or "no" to approve or cancel the job when using the terraform CLI. 

In the UI only member of that team defined in the template will be able to approved or cancel the job, in this case "TERRAFORM_CLI" won't exist because is something used internally only

What you can do is do the following.
```
POST {{API ENDPOINT}}/remote/tfe/v2/runs/{{JOB ID}}/actions/discard
Authorization: bearer {{PAT TOKEN}}
```

you can find the information for pat token [here](https://docs.terrakube.io/user-guide/organizations/api-tokens)

Another option is you can go to the database and change the job status as "cancelled"

**Comment by igorbrites** [comment]
@alfespa17 I've updated all the lock files, but I saw the same error again:

![image](https://github.com/AzBuilder/terrakube/assets/1479917/6292dd2a-6b5b-4998-af6b-e587a5b2febd)

During the plan, I saw that S3 error, but with a `.tfoutput` file:

```
 INFO 1 --- [io-8080-exec-10] o.t.a.p.s.c.TerraformOutputController    : Reading output from storage
 INFO 1 --- [io-8080-exec-10] o.t.a.p.s.aws.AwsStorageTypeServiceImpl  : Searching: tfoutput/6ccdd710-<<REDACTED>>-2e7eb8d18507/552/1c5e08a3-<<REDACTED>>-47390b02d97d.tfoutput
ERROR 1 --- [io-8080-exec-10] o.t.a.p.s.aws.AwsStorageTypeServiceImpl  : S3 Not found: The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: <<REDACTED>>; S3 Extended Request ID: 2wIzo7R5pj<<REDACTED>>a4HaJQ=; Proxy: null)
```

The file indeed doesn't exist in that folder:

![image](https://github.com/AzBuilder/terrakube/assets/1479917/1d24f0dd-2aac-49cd-a542-1b4cfcf5f98e)

To give you a context, I'm not using the VCS settings. On my pipelines, I plan and apply each folder using the TF commands directly (I saw on another issue that mixing the configs can be tricky).

So, I always run a plan and an apply directly on the folder. I have no more clues to understand what can be done in this case. I hope my info can help you somehow.

Thanks!

**Comment by alfespa17** [comment]
So If I understand correctly you have everything inside a GIT repository but you are running the Terraform plan/apply directly in the folder using GitHub actions.

Do you have a ".terraform.lock.hcl" file or ".terraform" folder inside the git repository folder that you are using?

If you have ".terraform.lock.hcl" file or ".terraform" there, could you delete those from your git folder and trigger the gihub action again to run the plan/apply?

**Comment by alfespa17** [comment]
When you run the CLI-driven workflow, terraform upload all the content of the folder where you are running the terraform command in a tar.gz file like this

![image](https://github.com/AzBuilder/terrakube/assets/4461895/ab2e66bf-e978-47fa-b51a-f27b66aa424d)

Maybe there is some issue with the ".terraform folder" or the ".terraform.lock.hcl file", if you have those in your GIT repository but I am just guessing right now.

Terrakube will always run a terraform init before the plan or apply internally like this, maybe that can fix your issue but I am just guessing now because I have no way to reproduce it
![image](https://github.com/AzBuilder/terrakube/assets/4461895/b683c2c4-c047-42a2-a55b-82d2e3ddb4fe)

**Comment by igorbrites** [comment]
> So If I understand correctly you have everything inside a GIT repository but you are running the Terraform plan/apply directly in the folder using GitHub actions.

Exactly.

> Do you have a ".terraform.lock.hcl" file or ".terraform" folder inside the git repository folder that you are using?

I commit the lock file following [terraform's recommendation](https://bit.ly/475U6mp), but the `.terraform` folder is generated during the `terraform init` command I run before everything, I don't commit it.

> If you have ".terraform.lock.hcl" file or ".terraform" there, could you delete those from your git folder and trigger the gihub action again to run the plan/apply?

If I run the plan without the `.terraform` folder, it gives me this error:

```
> terraform plan
╷
│ Error: Backend initialization required, please run "terraform init"
│ 
│ Reason: Initial configuration of the requested backend "remote"
```

**Comment by alfespa17** [comment]
Hello.

Can you do the following request:

```
GET {{terrakubeApi}}/api/v1/organization/{{organizationId}}/workspace/{{workspaceId}}
Authorization: Bearer {{PAT TOKEN}}
```

This will give you the workspace information, you can get the terraform files that the workspace is using when running the plan/apply from the property "source" and the UUID in the value will be the "contentId".

You can get the file form your storage in the following path in your S3 bucket

```
"content/{{contentId}}/terraformContent.tar.gz"
```

Double check if the workspace is using the correct files there.

Terrakube will use those files remotely to run the following when running the plan:

1 - Run terraform init with an override for the backend to use the S3 storage
2 - Run terraform plan and save the plan to file in the bucket 

Now we need to get the terraform plan file 

Do the following request

```
GET {{terrakubeApi}}/api/v1/organization/{{organizationId}}/job/{{jobId}}?include=step
```

From the response check the step information and look for stepNumber=100 that will give you the stepIf to get the terraform plan file.

You will find the plan file in the following path:

```
"tfstate/" + organizationId + "/" + workspaceId + "/" + jobId + "/" + stepId + "/" + terraformLibrary.tfPlan;
```

The remote apply in terrakube will basically do the following:

1 - Download the terraformContent.tar.gz again
2 - Run terraform init
3 - Run terraform apply using the plan file something like: 
```
terraform apply -input=false terraformLibrary.tfPlan
```

Maybe doing that you can at least reproduce the same error locally and get more information about what could be causing the issue
